File: 2024/albert-arvidsson-polyglot/day_01/part_1/solve.rb
#!/usr/bin/env ruby

require 'bundler/inline'

gemfile do
  source 'https://rubygems.org'
  gem 'debug'
  gem 'fiddle'
end

def solve(line)
  line
end

def parse_line(line)
  line.scan(/\d+/).map(&:to_i)
end

def solve_line(line)
  line.sum
end

File.readlines(ARGV[0]).map(&:strip)
    .map { |line| parse_line(line) }
    .transpose
    .map(&:sort)
    .transpose
    .map { |(a, b)| (a.abs - b.abs).abs }
    .sum
    .then(&method(:puts))

File: 2024/albert-arvidsson-polyglot/day_03/part_2/solve.exs
#!/usr/bin/env elixir

product = fn x, y -> String.to_integer(x) * String.to_integer(y) end

~r/mul\((\d{1,3}),(\d{1,3})\)|do\(\)|don't\(\)/
|> Regex.scan(System.argv() |> hd |> File.read!())
|> Enum.reduce({0, true}, fn
  ["do()"], {sum, _} -> {sum, true}
  ["don't()"], {sum, _} -> {sum, false}
  ["mul(" <> _, _, _], {sum, false} -> {sum, false}
  [_, x, y], {sum, _} -> {sum + product.(x, y), true}
end)
|> elem(0)
|> IO.puts()
